File: support/cpp/lib/coda/runtime/typeregistry.cpp
// ============================================================================
// Coda Type Registry
// ============================================================================

#include "coda/runtime/typeregistry.h"
#include "coda/descriptors.h"

#include <assert.h>

namespace coda {
namespace runtime {

using namespace coda::descriptors;

TypeRegistry &TypeRegistry::getInstance() {
  static TypeRegistry instance;
  return instance;
}

TypeRegistry& TypeRegistry::addSubtype(const StructType* subtype) {
  assert(subtype->getTypeId() > 0);
  const StructType* base = subtype->getBaseType();
  assert(base != NULL);
  while (base->getBaseType() != NULL) {
    base = base->getBaseType();
  }
  subtype_map::iterator si = subtypes.find(base);
  if (si == subtypes.end()) {
    si = subtypes.insert(std::make_pair(base, typeid_map())).first;
  }
  typeid_map& subtypesForBase = si->second;
  typeid_map::iterator it = subtypesForBase.find(subtype->getTypeId());
  if (it != subtypesForBase.end()) {
    throw std::runtime_error(
        "Error registering type " + subtype->getFullName() + ": subtype ID already registered.");
  }
  subtypesForBase[subtype->getTypeId()] = subtype;
  assert(getSubtype(base, subtype->getTypeId()) == subtype);
  return *this;
}

const coda::descriptors::StructType* TypeRegistry::getSubtype(
    const StructType* base, int32_t typeId) const {
  assert(base->getBaseType() == NULL);
  subtype_map::const_iterator it = subtypes.find(base);
  if (it == subtypes.end()) {
    return NULL;
  }
  const typeid_map& typeIds = it->second;
  typeid_map::const_iterator ti = typeIds.find(typeId);
  if (ti == typeIds.end()) {
    return NULL;
  }
  return ti->second;
}

//const typeid_map& TypeRegistry::getSubtypes(const StructType* base) const {
//  return self.__subtypes.get(id(base), {})
//}

TypeRegistry& TypeRegistry::addFile(const FileDescriptor* file) {
  const std::vector<StructType*>& structs = file->getStructs();
  for (std::vector<StructType*>::const_iterator it = structs.begin(), itEnd = structs.end();
      it != itEnd; ++it) {
    addStruct(*it);
  }
  return *this;
}

void TypeRegistry::addStruct(const StructType* st) {
  if (st->getBaseType() != NULL) {
    addSubtype(st);
  }
  const std::vector<StructType*>& structs = st->getStructs();
  for (std::vector<StructType*>::const_iterator it = structs.begin(), itEnd = structs.end();
      it != itEnd; ++it) {
    addStruct(*it);
  }
}

}} // namespace

File: libs/coda/runtime/typeregistry.py
'''Registry for extensions and subclasses.'''

class TypeRegistry:
  def __init__(self):
    self.__subtypes = {}
    self.__extensions = {}

  def addSubtype(self, subtype):
    '''Register a type as being a subtype of a given base type.'''
    assert subtype.getTypeId() is not None, subtype.getName()
    base = subtype.getBaseType()
    assert base is not None
    while base.getBaseType():
      base = base.getBaseType()
    try:
      subtypesForBase = self.__subtypes[id(base)]
    except KeyError:
      subtypesForBase = self.__subtypes[id(base)] = {}
    if subtype.getTypeId() in subtypesForBase:
      raise AssertionError(
          "Error registering type {0}: subtype ID {1} already registered".\
              formatter(subtype.getFullName(), subtype.getTypeId()))
    subtypesForBase[subtype.getTypeId()] = subtype
    return self

  def getSubtype(self, base, typeId):
    '''Retrieve a subtype of a base type by subtype ID.'''
    subtypesForBase = self.__subtypes.get(id(base))
    if subtypesForBase:
      return subtypesForBase.get(typeId)
    return None

  def getSubtypes(self, base):
    '''Retrieve all subtype of a base type.'''
    return self.__subtypes.get(id(base), {})

  def addFile(self, file):
    '''Add all subtypes and extensions registered within a file.'''
    def addStruct(struct):
      if struct.getBaseType() is not None:
        self.addSubtype(struct)
      for st in struct.getStructs():
        addStruct(st)
    for struct in file.getStructs():
      addStruct(struct)

  def getExtension(self, struct, fieldId):
    return self.__extensions.get(id(struct), {}).get(fieldId)

  def addExtension(self, extField):
    try:
      extensionsForStruct = self.__extensions[id(extField.getExtends())]
    except KeyError:
      extensionsForStruct = self.__subtypes[id(extField.getExtends())] = {}
    assert extField.getId() not in extensionsForStruct, \
        'Duplicate extension id for struct ' + extField.getExtends().getName()
    extensionsForStruct[extField.getId()] = extField

  INSTANCE = None

TypeRegistry.INSTANCE = TypeRegistry()
